### PR TITLE
Use GetFolderPath(UserProfile) to obtain HOME dir

### PR DESCRIPTION
**Pull request by am11**
The BCL API `Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)` provides fallback in case environment variable, corresponding to platform-specific home directory, is not set. e.g. on Unix it fallback to `getpwuid_r` call and similarly on Windows it uses a Win32 API as fallback.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs [hunk 1]
```diff
@@ -493,10 +493,9 @@ private static bool IsDevelopmentCertificate(X509Certificate2 certificate)
 
     private bool TryGetCertificatePath([NotNullWhen(true)] out string? path)
     {
-        // This will go away when we implement
-        // https://github.com/aspnet/Hosting/issues/1294
+        // See https://github.com/aspnet/Hosting/issues/1294
```
Comment (am11) [inline src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:496]: This request was closed as "Won't fix (at this time)".

**Comment by japui** [comment]
GetFolderPath(UserProfile) para obter o diretório HOME

File: src/Logging.AzureAppServices/src/WebAppContext.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ internal sealed class WebAppContext : IWebAppContext
     private WebAppContext() { }
 
     /// <inheritdoc />
-    public string HomeFolder { get; } = Environment.GetEnvironmentVariable("HOME");
+    public string HomeFolder { get; } = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
```
Comment (adityamandaleeka) [inline src/Logging.AzureAppServices/src/WebAppContext.cs:21]: This is not correct. The "HOME" environment variable has a special meaning in App Service where this is used: https://docs.microsoft.com/en-us/azure/app-service/reference-app-settings?tabs=kudu%2Cdotnet

**Comment by adityamandaleeka** [comment]
@am11 Does this change fix any issues you're seeing?

The App Service logging change is incorrect (see my other comment) and the other ones don't seem like they change anything valuable.

File: src/Logging.AzureAppServices/src/WebAppContext.cs (hunk 2 above)
Comment (am11) [inline src/Logging.AzureAppServices/src/WebAppContext.cs:21]:
```suggestion
    public string HomeFolder { get; } = Environment.GetEnvironmentVariable("HOME");
```

**Comment by am11** [comment]
@adityamandaleeka, when running web apps in environment where HOME is not set (linux container), kerstrel was not picking up certificates before this change. I have switched to GetFolderPath(UserPath), which provides appropriate fallback for each operating system supported by the runtime.

File: src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs [hunk 3]
```diff
@@ -91,8 +89,8 @@ private DefaultKeyStorageDirectories()
         // the %HOME% variable to build up our base key storage path.
         if (!String.IsNullOrEmpty(Environment.GetEnvironmentVariable("WEBSITE_INSTANCE_ID")))
         {
-            var homeEnvVar = Environment.GetEnvironmentVariable("HOME");
-            if (!String.IsNullOrEmpty(homeEnvVar))
+            var homeEnvVar = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
```
Comment (adityamandaleeka) [inline src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs:92]: I think this is another case where %HOME% may not be the same as this special folder (this is specific to the Azure Websites case).

File: src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs [hunk 4]
```diff
@@ -49,14 +48,13 @@ private DefaultKeyStorageDirectories()
         {
             retVal = GetKeyStorageDirectoryFromBaseAppDataPath(localAppDataFromEnvPath);
         }
-        else if (userProfilePath != null)
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && homePath != null)
         {
-            retVal = GetKeyStorageDirectoryFromBaseAppDataPath(Path.Combine(userProfilePath, "AppData", "Local"));
+            retVal = GetKeyStorageDirectoryFromBaseAppDataPath(Path.Combine(homePath, "AppData", "Local"));
```
Comment (adityamandaleeka) [inline src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs:53]: Is this any different from using Environment.SpecialFolder.LocalApplicationData at this point?

File: src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs [hunk 5]
```diff
@@ -49,14 +48,13 @@ private DefaultKeyStorageDirectories()
         {
             retVal = GetKeyStorageDirectoryFromBaseAppDataPath(localAppDataFromEnvPath);
         }
-        else if (userProfilePath != null)
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && homePath != null)
```
Comment (adityamandaleeka) [inline src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs:51]: Factor out the homePath != null:
```c#
        else if (homePath != null)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
               ...
            }
            else
            {
               ...
            }
        }
```

**Comment by adityamandaleeka** [comment]
I left some more comments. I'd feel better if the DataProtection changes were limited to only affect non-Windows codepaths (which is what you're trying to fix anyway). It's clear from the code/comments that there's some history and specific decisions were made around how (and in what order) to probe for the directories, and in some cases %HOME% means something different from the UserProfile directory.

File: src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs (hunk 4 above)
Comment (am11) [reply src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs]: I have changed it to `Environment.SpecialFolder.LocalApplicationData`.

Comment (adityamandaleeka) [reply src/DataProtection/DataProtection/src/Repositories/DefaultKeyStorageDirectories.cs]: But `Environment.SpecialFolder.LocalApplicationData` is already checked first above so I don't think this makes sense.

**Comment by adityamandaleeka** [comment]
@am11 Thanks for the update. As I said above, I'd still prefer this not to mess with the existing probing order on Windows for the data protection code.

I'm also curious if, on Linux, the Environment.SpecialFolder.LocalApplicationData fallback (which is currently taken after the $HOME check) will ever be used anymore after your change, and whether anyone would currently be depending on that.

cc: @wtgodbe

**Comment by am11** [comment]
> not to mess with the existing probing order on Windows

I don't follow, the ordering hasn't changed on Windows?

**Review by adityamandaleeka — APPROVED** [review]

**Pull request merged**